include/pybind11/detail/smart_holder_type_casters.h

```diff
@@ -616,13 +616,39 @@ struct smart_holder_type_caster_load {
                               "instance cannot safely be transferred to C++.");
         }
 
+        // Default constructed temporary variable to store the extracted deleter in.
+        D extracted_deleter;
```

Comment:
Can we guarantee that D is always default construct-able?

Reply:
We cannot, but that is a requirement of the current code, a bit below this line we used to have:
https://github.com/pybind/pybind11/blob/edfaaed87ea3a869eaa1947f3d40cb62958295ca/include/pybind11/detail/smart_holder_type_casters.h#L625

Which is the constructor for the unique pointer where the deleter will be default constructed; [cppreference](https://en.cppreference.com/w/cpp/memory/unique_ptr/unique_ptr#:~:text=the%20stored%20deleter.-,Requires%20that%20Deleter%20is%20DefaultConstructible%20and%20that%20construction%20does%20not%20throw%20an%20exception.,-This%20overload%20participates).

I personally don't think this is a problem, overcoming the limitation is probably possible, but would make the code more convoluted.

Reply:
Marking as resolved, this requirement is no longer present.